src/app/CommandHandler.cpp

```diff
@@ -108,13 +110,20 @@ CHIP_ERROR CommandHandler::ProcessCommandDataElement(CommandDataElement::Parser
 exit:
     if (err != CHIP_NO_ERROR)
     {
-        // The Path is not present when there is an error to be conveyed back. ResponseCommandElement would only have status code,
-        // set the error with CHIP_NO_ERROR, then continue to process rest of commands
-        AddStatusCode(nullptr, GeneralStatusCode::kInvalidArgument, Protocols::SecureChannel::Id,
-                      Protocols::SecureChannel::kProtocolCodeGeneralFailure);
-        err = CHIP_NO_ERROR;
+        chip::app::CommandPathParams returnStatusParam = { endpointId,
+                                                           0, // GroupId
+                                                           clusterId, commandId, (chip::app::CommandPathFlags::kEndpointIdValid) };
+
+        // The Path is the path in the request if there are any error occurred before we dispatch the command to clusters.
+        // Currently, it could be failed to decode Path or failed to find cluster / command on desired endpoint.
+        // TODO: The behavior when receiving a malformed message is not clear in the Spec. (Spec#3259)
+        AddStatusCode(&returnStatusParam,
+                      err == CHIP_ERROR_INVALID_PROFILE_ID ? GeneralStatusCode::kNotFound : GeneralStatusCode::kInvalidArgument,
+                      Protocols::SecureChannel::Id, Protocols::SecureChannel::kProtocolCodeGeneralFailure);
     }
-    return err;
+    // We have handled the error status above and puts the error status in response, now return success status so we can process
+    // other commands in the invoke request (by caller).
```

Comment:
```suggestion
    // We have handled the error status above and put the error status in response, now return success status so we can process
    // other commands in the invoke request.
```
Thank you, this is much clearer now.